# 4

In [2]:
import pandas as pd
import numpy as np
import wooldridge as woo
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import statsmodels.formula.api as smf
import statsmodels.stats.api as sms
import scipy.stats
import numpy.ma as ma
import patsy as pt
import statsmodels.stats.outliers_influence as smo
import statsmodels.api as sm
from simple_colors import *
from RegscorePy import mallow
from sklearn.model_selection import KFold
import warnings 

## PART 1

### Use the data in DISCRIM from wooldridge to answer this question.

In [3]:
disc = woo.data("DISCRIM")
disc.head()

psoda  pfries  pentree  wagest  nmgrs  nregs  hrsopen   emp  psoda2  \
0   1.12    1.06     1.02    4.25    3.0    5.0     16.0  27.5    1.11   
1   1.06    0.91     0.95    4.75    3.0    3.0     16.5  21.5    1.05   
2   1.06    0.91     0.98    4.25    3.0    5.0     18.0  30.0    1.05   
3   1.12    1.02     1.06    5.00    4.0    5.0     16.0  27.5    1.15   
4   1.12     NaN     0.49    5.00    3.0    3.0     16.0   5.0    1.04   

   pfries2  ...  county    lpsoda   lpfries    lhseval    lincome  ldensity  \
0     1.11  ...      18  0.113329  0.058269  11.906993  10.704008  8.301521   
1     0.89  ...      18  0.058269 -0.094311  11.906993  10.704008  8.301521   
2     0.94  ...      12  0.058269 -0.094311  12.038836  10.625319  9.341369   
3     1.05  ...      10  0.113329  0.019803  12.052921  10.827071  9.029418   
4     1.01  ...      10  0.113329       NaN  12.425610  11.188399  6.579251   

   NJ  BK  KFC  RR  
0   1   0    0   1  
1   1   1    0   0  
2   1   1    0   0  
3   1   0    0   1  
4   1   1    0   0  

[5 rows x 37 columns]

In [4]:
disc.describe()

psoda      pfries     pentree      wagest       nmgrs       nregs  \
count  402.000000  393.000000  398.000000  390.000000  404.000000  388.000000   
mean     1.044876    0.921985    1.322186    4.615641    3.420297    3.608247   
std      0.088687    0.105881    0.643084    0.347015    1.018408    1.243540   
min      0.730000    0.670000    0.490000    4.250000    1.000000    1.000000   
25%      0.980000    0.850000    0.950000    4.250000    3.000000    3.000000   
50%      1.060000    0.930000    1.020000    4.500000    3.000000    3.000000   
75%      1.085000    1.000000    1.470000    4.950000    4.000000    4.000000   
max      1.490000    1.270000    3.950000    5.750000   10.000000    8.000000   

          hrsopen         emp      psoda2     pfries2  ...      county  \
count  410.000000  404.000000  388.000000  382.000000  ...  410.000000   
mean    14.439024   17.621906    1.044948    0.941230  ...   13.658537   
std      2.809987    9.423264    0.093567    0.109304  ...    8.045439   
min      7.000000    3.000000    0.410000    0.690000  ...    1.000000   
25%     12.000000   11.375000    1.000000    0.840000  ...    6.000000   
50%     15.500000   16.375000    1.050000    0.940000  ...   14.000000   
75%     16.000000   21.000000    1.102500    1.010000  ...   20.000000   
max     24.000000   80.000000    1.400000    1.370000  ...   29.000000   

           lpsoda     lpfries     lhseval     lincome    ldensity          NJ  \
count  402.000000  393.000000  409.000000  409.000000  409.000000  410.000000   
mean     0.040321   -0.087813   11.828533   10.719937    7.958750    0.807317   
std      0.084730    0.115152    0.389349    0.284479    0.995858    0.394888   
min     -0.314711   -0.400478   10.431170    9.675268    5.093750    0.000000   
25%     -0.020203   -0.162519   11.588960   10.542257    7.418181    1.000000   
50%      0.058269   -0.072571   11.865693   10.742292    7.961370    1.000000   
75%      0.081548    0.000000   12.082774   10.914743    8.641179    1.000000   
max      0.398776    0.239017   13.067696   11.824292   10.631929    1.000000   

               BK         KFC          RR  
count  410.000000  410.000000  410.000000  
mean     0.417073    0.195122    0.241463  
std      0.493678    0.396779    0.428493  
min      0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000  
50%      0.000000    0.000000    0.000000  
75%      1.000000    0.000000    0.000000  
max      1.000000    1.000000    1.000000  

[8 rows x 37 columns]

In [5]:
disc.isnull().sum()

psoda        8
pfries      17
pentree     12
wagest      20
nmgrs        6
nregs       22
hrsopen      0
emp          6
psoda2      22
pfries2     28
pentree2    24
wagest2     21
nmgrs2       6
nregs2      22
hrsopen2    11
emp2        13
compown      0
chain        0
density      1
crmrte       1
state        0
prpblck      1
prppov       1
prpncar      1
hseval       1
nstores      0
income       1
county       0
lpsoda       8
lpfries     17
lhseval      1
lincome      1
ldensity     1
NJ           0
BK           0
KFC          0
RR           0
dtype: int64

### Estimate the regression model $ log(psoda) = β_{0} + β_{1}prpblck + β_{2}log(income) + β_{3}prppov + u $  and report the results in the usual form. Is $ β_{1} $ statistically different from zero at the 5% level against a two-sided alternative? What about at the 1% level?


In [6]:
result = smf.ols('np.log(psoda) ~ prpblck + np.log(income) + prppov', disc).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(psoda)   R-squared:                       0.087
Model:                            OLS   Adj. R-squared:                  0.080
Method:                 Least Squares   F-statistic:                     12.60
Date:                Fri, 18 Nov 2022   Prob (F-statistic):           6.92e-08
Time:                        18:50:44   Log-Likelihood:                 439.04
No. Observations:                 401   AIC:                            -870.1
Df Residuals:                     397   BIC:                            -854.1
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         -1.4633      0.294     -4.982      0.000      -2.041      -0.886
prpblck            0.0728      0.031      2.373      0.018       0.013       0.133
np.log(income)     0.1370      0.027      5.119      0.000       0.084       0.190
prppov             0.3804      0.133      2.864      0.004       0.119       0.641
==============================================================================
Omnibus:                       12.002   Durbin-Watson:                   1.727
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               24.056
Skew:                          -0.014   Prob(JB):                     5.97e-06
Kurtosis:                       4.200   Cond. No.                         834.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The p-value for $ B_{1} $ is 0.018. Therefore, $ B_{1} $ is statistically different from 0 at the 5% level, yet not at the 1% level.

### What is the correlation between $ log(income) $ and $ prppov $? VIF? Is each variable statistically significant in any case? Report the two-sided p-values.

In [7]:
print(ma.corrcoef(ma.masked_invalid(np.log(disc.income)), ma.masked_invalid(disc.prppov)))

[[1.0 -0.8384669733579039]
 [-0.8384669733579039 1.0]]


The correlation coefficient between $ log(income) $ and $ prppov $ is -0.838, suggesting multicollinearity between the two variables. 

In [8]:
y, X = pt.dmatrices('np.log(psoda) ~ prpblck + np.log(income) + prppov',
                    data=disc, return_type='dataframe')

K = X.shape[1]
VIF = np.empty(K)
for i in range(K):
    VIF[i] = smo.variance_inflation_factor(X.values, i)
print(f'VIF: \n{VIF}\n')

VIF: 
[5.22495302e+03 1.92215952e+00 3.51872083e+00 4.91521981e+00]



Moreover, $ prppov $ is beyond the VIF treshold set at 4, suggesting again multicollinearity.

In [9]:
print("The p-value for log(income) is:", result.pvalues[2])
print("The p-value for prppov is:", result.pvalues[3])

The p-value for log(income) is: 4.802040557260267e-07
The p-value for prppov is: 0.004400360031387504


However, the p-values for the two variables suggest that both $ log(income) $ and $ prppov $ are statistically significant.

### To the regression in part (a), add the variable log(hseval). Interpret its coefficient and report the two-sided p-value for $ H_{0} : βlog(hseval) = 0 $.

In [10]:
result2 = smf.ols('np.log(psoda) ~ prpblck + np.log(income) + prppov + np.log(hseval)', disc).fit()
result2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(psoda)   R-squared:                       0.184
Model:                            OLS   Adj. R-squared:                  0.176
Method:                 Least Squares   F-statistic:                     22.31
Date:                Fri, 18 Nov 2022   Prob (F-statistic):           1.24e-16
Time:                        18:50:44   Log-Likelihood:                 461.55
No. Observations:                 401   AIC:                            -913.1
Df Residuals:                     396   BIC:                            -893.1
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         -0.8415      0.292     -2.878      0.004      -1.416      -0.267
prpblck            0.0976      0.029      3.334      0.001       0.040       0.155
np.log(income)    -0.0530      0.038     -1.412      0.159      -0.127       0.021
prppov             0.0521      0.134      0.388      0.699      -0.212       0.317
np.log(hseval)     0.1213      0.018      6.860      0.000       0.087       0.156
==============================================================================
Omnibus:                       16.452   Durbin-Watson:                   1.973
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               40.077
Skew:                          -0.016   Prob(JB):                     1.98e-09
Kurtosis:                       4.548   Cond. No.                     1.31e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.31e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [11]:
print("The p-value for log(hseval) is:", result2.pvalues[4])

The p-value for log(hseval) is: 2.6681248076804145e-11


The p-value for $ log(hseval) $ suggests that the variable is statistically significant.

An increase of 1% in housing medial value increases the price of a medium soda by around 0.1213% - when the other variables are fixed.

### In the regression in part (c), what happens to the individual statistical significance of $ log(income) $ and $ prppov $? Are these variables jointly significant? (Compute a p-value.) What do you make of your answers?

In [12]:
print("The new p-value for log(income) is:", result2.pvalues[2])
print("The new p-value for prppov is:", result2.pvalues[3])

The new p-value for log(income) is: 0.1587064623592347
The new p-value for prppov is: 0.6985706123569336


Both $ log(income) $ and $ prppov $ are not statistically significant in the new model.

In [13]:
hypotheses = ["np.log(income) = 0", "prppov = 0"]
result2.f_test(hypotheses) 

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=3.522687204208135, p=0.030448537491840212, df_denom=396, df_num=2>

However, as shown by the F-test, the two variables are jointly significant at the 5% level.

This can happen because $ log(income) $ and $ prppov $ are multicorrelated; they are individually insignificant but jointly significant.

### Given the results of the previous regressions, which one would you report as most reliable in determining whether the racial makeup of a zip code influences local fast-food prices?

Given the result of both regression, I would expect model 2 to be more reliable in predicting the influence of racial makeup in fast-food prices. As a matter of fact, the second model contains more variables to explain the y-variable and has a better R squared. Even thought $ log(income) $ and $ prppov $ are multicorrelated and unsignificant, this should not be a concern for our prediction, also becuase they are jointly significant. 

In [14]:
print("AIC for model 1 is: ", result.aic)
print("AIC for model 2 is: ", result2.aic)

AIC for model 1 is:  -870.0718830241235
AIC for model 2 is:  -913.0952568304301


In [15]:
print("BIC for model 1 is: ", result.bic)
print("BIC for model 2 is: ", result2.bic)

BIC for model 1 is:  -854.0960373148972
BIC for model 2 is:  -893.1254496938973


Both AIC and BIC confirm that model 2 is the best model.

## PART 2

### Use the VOTE1 dataset from theWooldridge python module to answer the following question. Import this data into your notebook

In [16]:
vote = woo.data("VOTE1")
vote.head()

state  district  democA  voteA     expendA     expendB  prtystrA  lexpendA  \
0    AL         7       1     68  328.295990    8.737000        41  5.793916   
1    AK         1       0     62  626.377014  402.476990        60  6.439952   
2    AZ         2       1     73   99.607002    3.065000        55  4.601233   
3    AZ         3       0     69  319.690002   26.281000        64  5.767352   
4    AR         3       0     75  159.220993   60.054001        66  5.070293   

   lexpendB     shareA  
0  2.167567  97.407669  
1  5.997638  60.881039  
2  1.120048  97.014763  
3  3.268846  92.403702  
4  4.095244  72.612473

In [17]:
vote.describe()

district      democA       voteA      expendA      expendB  \
count  173.000000  173.000000  173.000000   173.000000   173.000000   
mean     8.838150    0.554913   50.502890   310.611005   305.088537   
std      8.768823    0.498418   16.784761   280.985381   306.278339   
min      1.000000    0.000000   16.000000     0.302000     0.930000   
25%      3.000000    0.000000   36.000000    81.634003    60.054001   
50%      6.000000    1.000000   50.000000   242.781998   221.529999   
75%     11.000000    1.000000   65.000000   457.410004   450.716003   
max     42.000000    1.000000   84.000000  1470.673950  1548.192993   

         prtystrA    lexpendA    lexpendB      shareA  
count  173.000000  173.000000  173.000000  173.000000  
mean    49.757225    5.025556    4.944369   51.076545  
std      9.983650    1.601602    1.571143   33.483575  
min     22.000000   -1.197328   -0.072571    0.094635  
25%     44.000000    4.402246    4.095244   18.867996  
50%     50.000000    5.492164    5.400558   50.849903  
75%     56.000000    6.125580    6.110837   84.255096  
max     71.000000    7.293476    7.344844   99.495003

In [18]:
vote.isnull().sum()

state       0
district    0
democA      0
voteA       0
expendA     0
expendB     0
prtystrA    0
lexpendA    0
lexpendB    0
shareA      0
dtype: int64

### Estimate the following model: $ voteA = β_{0} + β_{1}prtystrA + β_{2}log(expendA) + β_{3}democA + β_{4}log(expendB) + u $ <br> Obtain the OLS residuals, $ \hat{u}_{i} $, and regress these on all of the independent variables. Explain why you obtain $ R^2 $ = 0.

In [19]:
fit = smf.ols('voteA ~ prtystrA + np.log(expendA) + democA + np.log(expendB)', vote).fit()
fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.801
Model:                            OLS   Adj. R-squared:                  0.796
Method:                 Least Squares   F-statistic:                     169.2
Date:                Fri, 18 Nov 2022   Prob (F-statistic):           8.09e-58
Time:                        18:50:45   Log-Likelihood:                -593.20
No. Observations:                 173   AIC:                             1196.
Df Residuals:                     168   BIC:                             1212.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          37.6614      4.736      7.952      0.000      28.312      47.011
prtystrA            0.2519      0.071      3.534      0.001       0.111       0.393
np.log(expendA)     5.7793      0.392     14.750      0.000       5.006       6.553
democA              3.7929      1.407      2.697      0.008       1.016       6.570
np.log(expendB)    -6.2378      0.397    -15.694      0.000      -7.022      -5.453
==============================================================================
Omnibus:                        6.304   Durbin-Watson:                   1.525
Prob(Omnibus):                  0.043   Jarque-Bera (JB):                6.030
Skew:                           0.448   Prob(JB):                       0.0491
Kurtosis:                       3.182   Cond. No.                         429.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [20]:
u_fit = smf.ols('fit.resid ~ prtystrA + np.log(expendA) + democA + np.log(expendB)', vote).fit()
u_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              fit.resid   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.024
Method:                 Least Squares   F-statistic:                     0.000
Date:                Fri, 18 Nov 2022   Prob (F-statistic):               1.00
Time:                        18:50:45   Log-Likelihood:                -593.20
No. Observations:                 173   AIC:                             1196.
Df Residuals:                     168   BIC:                             1212.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept       -1.465e-14      4.736  -3.09e-15      1.000      -9.350       9.350
prtystrA         -1.11e-16      0.071  -1.56e-15      1.000      -0.141       0.141
np.log(expendA)  8.771e-15      0.392   2.24e-14      1.000      -0.774       0.774
democA           2.665e-15      1.407   1.89e-15      1.000      -2.777       2.777
np.log(expendB)  4.441e-16      0.397   1.12e-15      1.000      -0.785       0.785
==============================================================================
Omnibus:                        6.304   Durbin-Watson:                   1.525
Prob(Omnibus):                  0.043   Jarque-Bera (JB):                6.030
Skew:                           0.448   Prob(JB):                       0.0491
Kurtosis:                       3.182   Cond. No.                         429.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

We would expect a $ R^2 $ equals to 0 because, as the OLS assumption claims, errors in the sample should not be correlated to the independent variables. Therefore, the predictors should not be able to explain the error term (so we have $ R^2 $=0).

### Now, compute the Breusch-Pagan test for heteroskedasticity. Use the F statistic version and report the p-value.

In [21]:
name = ["Lagrange multiplier statistic", "p-value", "f-value", "f p-value"]
test = sms.het_breuschpagan(fit.resid, fit.model.exog)
print("BP Results:")
print(list(zip(name, test)))

BP Results:
[('Lagrange multiplier statistic', 9.093355957811157), ('p-value', 0.05880791685694346), ('f-value', 2.330112683716247), ('f p-value', 0.058057514088555584)]


As the f p-value is about 0.058, we fail to reject $ H_{0} $ and assume there might not be heteroskedasticity for alpha set at 5% and 1%. 

### Compute the special case of the White test for heteroskedasticity, again using the F statistic form. How strong is the evidence for heteroskedasticity now?

**Classic White test**

In [22]:
name = ["Lagrange multiplier statistic", "p-value", "f-value", "f p-value"]
test = sm.stats.diagnostic.het_white(fit.resid, fit.model.exog)
print("White Test Results:")
print(list(zip(name, test)))

White Test Results:
[('Lagrange multiplier statistic', 31.101516025646937), ('p-value', 0.0032582616087888055), ('f-value', 2.680757782482806), ('f p-value', 0.0019733250042552435)]


According to the classic White test, there is evidence of heteroskedasticity.

**Special case White test**

In [23]:
fit_sqr = smf.ols('np.square(fit.resid) ~ fit.fittedvalues + np.square(fit.fittedvalues)', vote).fit()
fit_sqr.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                             OLS Regression Results                             
================================================================================
Dep. Variable:     np.square(fit.resid)   R-squared:                       0.032
Model:                              OLS   Adj. R-squared:                  0.020
Method:                   Least Squares   F-statistic:                     2.786
Date:                  Fri, 18 Nov 2022   Prob (F-statistic):             0.0645
Time:                          18:50:47   Log-Likelihood:                -1005.6
No. Observations:                   173   AIC:                             2017.
Df Residuals:                       170   BIC:                             2027.
Df Model:                             2                                         
Covariance Type:              nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                     171.8584     53.142      3.234      0.001      66.955     276.762
fit.fittedvalues               -4.2637      2.167     -1.968      0.051      -8.540       0.013
np.square(fit.fittedvalues)     0.0357      0.021      1.682      0.094      -0.006       0.078
==============================================================================
Omnibus:                      136.013   Durbin-Watson:                   1.957
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1393.398
Skew:                           2.927   Prob(JB):                    2.68e-303
Kurtosis:                      15.611   Cond. No.                     2.71e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.71e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [24]:
hypotheses = ["fit.fittedvalues = 0", "np.square(fit.fittedvalues) = 0"]
fit_sqr.f_test(hypotheses)

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=2.785827373447078, p=0.06449619861136317, df_denom=170, df_num=2>

The F-test for this regression is the result for the special case for the White test - we are giving some restriction to the parameters. It tells us that there is even less evidence for heteroskedasticity than the BP test. We accept $ H_{0} $.

### Estimate two more models. For the first use robust standard errors (HC0) and for the second use the FGLS procedure. Note any changes in your standard errors, significance level, and coefficients.

**Model 1 - HC0**

In [25]:
HC0_fit = smf.ols('voteA ~ prtystrA + np.log(expendA) + democA + np.log(expendB)', vote).fit(cov_type = 'HC0')
HC0_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.801
Model:                            OLS   Adj. R-squared:                  0.796
Method:                 Least Squares   F-statistic:                     169.1
Date:                Fri, 18 Nov 2022   Prob (F-statistic):           8.58e-58
Time:                        18:50:48   Log-Likelihood:                -593.20
No. Observations:                 173   AIC:                             1196.
Df Residuals:                     168   BIC:                             1212.
Df Model:                           4                                         
Covariance Type:                  HC0                                         
===================================================================================
                      coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          37.6614      4.355      8.649      0.000      29.127      46.196
prtystrA            0.2519      0.065      3.870      0.000       0.124       0.380
np.log(expendA)     5.7793      0.525     11.000      0.000       4.750       6.809
democA              3.7929      1.431      2.651      0.008       0.988       6.598
np.log(expendB)    -6.2378      0.351    -17.773      0.000      -6.926      -5.550
==============================================================================
Omnibus:                        6.304   Durbin-Watson:                   1.525
Prob(Omnibus):                  0.043   Jarque-Bera (JB):                6.030
Skew:                           0.448   Prob(JB):                       0.0491
Kurtosis:                       3.182   Cond. No.                         429.
==============================================================================

Notes:
[1] Standard Errors are heteroscedasticity robust (HC0)
"""

In [26]:
# SE
print("The SE for the original model was:\n", fit.bse)
print("The SE for the HC0 model is:\n", HC0_fit.bse)

The SE for the original model was:
 Intercept          4.736036
prtystrA           0.071293
np.log(expendA)    0.391820
democA             1.406520
np.log(expendB)    0.397460
dtype: float64
The SE for the HC0 model is:
 Intercept          4.354614
prtystrA           0.065101
np.log(expendA)    0.525381
democA             1.431029
np.log(expendB)    0.350974
dtype: float64


SE went down for the intercept, $ prtystrA $, and $ log(expendB) $. However, it increased for $ log(expendA) $ and $ democA$.

In [27]:
# Significance
print("The p-values for the original model were:\n", fit.pvalues)
print("The z-values for the HC0 model are:\n", HC0_fit.pvalues)

The p-values for the original model were:
 Intercept          2.564412e-13
prtystrA           5.296255e-04
np.log(expendA)    4.029908e-32
democA             7.716801e-03
np.log(expendB)    9.342922e-35
dtype: float64
The z-values for the HC0 model are:
 Intercept          5.212495e-18
prtystrA           1.090056e-04
np.log(expendA)    3.813122e-28
democA             8.037236e-03
np.log(expendB)    1.145213e-70
dtype: float64


The coefficients remain statistically significant.

In [28]:
# Coefficients
print("The parameters for the original model were:\n", fit.params)
print("The parameters for the HC0 model are:\n", HC0_fit.params)

The parameters for the original model were:
 Intercept          37.661415
prtystrA            0.251918
np.log(expendA)     5.779294
democA              3.792944
np.log(expendB)    -6.237836
dtype: float64
The parameters for the HC0 model are:
 Intercept          37.661415
prtystrA            0.251918
np.log(expendA)     5.779294
democA              3.792944
np.log(expendB)    -6.237836
dtype: float64


It appears the coefficient did not change or hardly changed - due apporximation reasons maybe we do not see it.

**Model 2 - FGLS**

In [29]:
u2 = fit.resid**2

In [30]:
u2_fit = smf.ols('np.log(u2) ~ prtystrA + np.log(expendA) + democA + np.log(expendB)', vote).fit()
u2_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             np.log(u2)   R-squared:                       0.037
Model:                            OLS   Adj. R-squared:                  0.014
Method:                 Least Squares   F-statistic:                     1.633
Date:                Fri, 18 Nov 2022   Prob (F-statistic):              0.168
Time:                        18:50:52   Log-Likelihood:                -372.96
No. Observations:                 173   AIC:                             755.9
Df Residuals:                     168   BIC:                             771.7
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           4.9597      1.326      3.740      0.000       2.342       7.577
prtystrA           -0.0250      0.020     -1.251      0.213      -0.064       0.014
np.log(expendA)    -0.0358      0.110     -0.326      0.744      -0.252       0.181
democA              0.2127      0.394      0.540      0.590      -0.565       0.990
np.log(expendB)    -0.1728      0.111     -1.553      0.122      -0.392       0.047
==============================================================================
Omnibus:                       57.246   Durbin-Watson:                   1.931
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              137.908
Skew:                          -1.437   Prob(JB):                     1.13e-30
Kurtosis:                       6.298   Cond. No.                         429.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [31]:
u2_fitvalues = u2_fit.fittedvalues
vari = np.exp(u2_fitvalues)
w = 1/vari**2

In [32]:
wls_fit = smf.wls('voteA ~ prtystrA + np.log(expendA) + democA + np.log(expendB)', weights = w, data = vote).fit()
wls_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            WLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.776
Model:                            WLS   Adj. R-squared:                  0.771
Method:                 Least Squares   F-statistic:                     145.8
Date:                Fri, 18 Nov 2022   Prob (F-statistic):           1.56e-53
Time:                        18:50:54   Log-Likelihood:                -600.44
No. Observations:                 173   AIC:                             1211.
Df Residuals:                     168   BIC:                             1227.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          43.9099      4.915      8.934      0.000      34.207      53.612
prtystrA            0.1898      0.071      2.691      0.008       0.051       0.329
np.log(expendA)     6.1749      0.386     16.012      0.000       5.414       6.936
democA              2.4745      1.327      1.865      0.064      -0.145       5.094
np.log(expendB)    -7.0865      0.432    -16.398      0.000      -7.940      -6.233
==============================================================================
Omnibus:                        4.729   Durbin-Watson:                   1.567
Prob(Omnibus):                  0.094   Jarque-Bera (JB):                5.006
Skew:                           0.227   Prob(JB):                       0.0818
Kurtosis:                       3.699   Cond. No.                         548.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [33]:
# SE
print("The SE for the original model was:\n", fit.bse)
print("The SE for the FGLS model is:\n", wls_fit.bse)

The SE for the original model was:
 Intercept          4.736036
prtystrA           0.071293
np.log(expendA)    0.391820
democA             1.406520
np.log(expendB)    0.397460
dtype: float64
The SE for the FGLS model is:
 Intercept          4.914740
prtystrA           0.070516
np.log(expendA)    0.385641
democA             1.326701
np.log(expendB)    0.432148
dtype: float64


The SEs for the coefficients changed. There has been an increase for the SE of the intercept, and $ log(expendB) $. The other SEs decreased. 

In [34]:
# Significance
print("The p-values for the original model were:\n", fit.pvalues)
print("The p-values for the FGLS model are:\n", wls_fit.pvalues)

The p-values for the original model were:
 Intercept          2.564412e-13
prtystrA           5.296255e-04
np.log(expendA)    4.029908e-32
democA             7.716801e-03
np.log(expendB)    9.342922e-35
dtype: float64
The p-values for the FGLS model are:
 Intercept          7.041221e-16
prtystrA           7.837046e-03
np.log(expendA)    1.232256e-35
democA             6.389885e-02
np.log(expendB)    1.062321e-36
dtype: float64


The p-values are pretty much the same; however, in the FGLS model $ democA $ is not significant for alpha 0.05 or 0.01.

In [35]:
# Coefficients
print("The parameters for the original model were:\n", fit.params)
print("The parameters for the FGLS model are:\n", wls_fit.params)

The parameters for the original model were:
 Intercept          37.661415
prtystrA            0.251918
np.log(expendA)     5.779294
democA              3.792944
np.log(expendB)    -6.237836
dtype: float64
The parameters for the FGLS model are:
 Intercept          43.909854
prtystrA            0.189781
np.log(expendA)     6.174878
democA              2.474548
np.log(expendB)    -7.086457
dtype: float64


Differently from the HC0 model, the coefficients changed - even though the variation is not too large.